packages/firestore/src/api/user_data_converter.ts

```diff
@@ -530,51 +530,60 @@ export class UserDataConverter {
     } else if (value instanceof DocumentKeyReference) {
       return new RefValue(value.databaseId, value.key);
     } else if (value instanceof FieldValueImpl) {
-      if (value instanceof DeleteFieldValueImpl) {
-        if (context.dataSource === UserDataSource.MergeSet) {
-          return null;
-        } else if (context.dataSource === UserDataSource.Update) {
-          assert(
-            context.path == null || context.path.length > 0,
-            'FieldValue.delete() at the top level should have already' +
-              ' been handled.'
-          );
-          throw context.createError(
-            'FieldValue.delete() can only appear at the top level ' +
-              'of your update data'
-          );
-        } else {
-          // We shouldn't encounter delete sentinels for queries or non-merge set() calls.
-          throw context.createError(
-            'FieldValue.delete() can only be used with update() and set() with {merge:true}'
-          );
-        }
-      } else if (value instanceof ServerTimestampFieldValueImpl) {
-        if (!isWrite(context.dataSource)) {
-          throw context.createError(
-            'FieldValue.serverTimestamp() can only be used with set()' +
-              ' and update()'
-          );
-        }
-        if (context.path === null) {
-          throw context.createError(
-            'FieldValue.serverTimestamp() is not currently' +
-              ' supported inside arrays'
-          );
-        }
-        context.fieldTransforms.push(
-          new FieldTransform(context.path, ServerTimestampTransform.instance)
-        );
+      return this.parseSentinelFieldValue(value, context);
+    } else {
+      throw context.createError(
+        `Unsupported field value: ${valueDescription(value)}`
+      );
+    }
+  }
+
+  private parseSentinelFieldValue(
+    value: FieldValueImpl,
+    context: ParseContext
+  ): FieldValue | null {
+    // Sentinels are only supported with writes, and not within arrays.
+    if (!isWrite(context.dataSource)) {
+      throw context.createError(
+        `${value.methodName} can only be used with update() and set().`
```

Comment:
Nit: Remove period to match other errors.

Reply:
Done.